fvm/environment/unsafe_random_generator.go

```diff
@@ -76,14 +79,24 @@ func (gen *unsafeRandomGenerator) seed() {
 		// header ID. The random number generator will be used by the
 		// UnsafeRandom function.
 		id := gen.blockHeader.ID()
-		source := rand.NewSource(int64(binary.BigEndian.Uint64(id[:])))
-		gen.rng = rand.New(source)
+		// extract the entropy from `id` and expand it into the required seed
+		hkdf := hkdf.New(func() hash.Hash { return sha256.New() }, id[:], nil, nil)
```

Comment:
txn index as the salt?

Reply:
Yep the idea is to include the transaction index somewhere in the PRG construction (as a KDF salt, as a PRG customizer..). 
I only wanted this PR to remove `math/rand` and use a safer PRG. Wiring the transaction index down the stack till the generator could be done in another PR.

Reply:
sgtm.  add a todo as a reminder

Reply:
@bluesign good questions!

> Is block ID predictable ? or can it be influenced?

Block ID can be biased by a consensus node. It is possible for consensus nodes to loop over possible blocks and choose one with a block ID with the random outputs they prefer. 
The protocol has a safer entropy source which is the random beacon, that is unpredictable and unbiasable (based on BLS threshold signatures). The plan is to wire its output to Cadence to export a safe randomness. 

> why KDF necessary here ?

Here we have an entropy source as an input (block ID + soon a transaction index) and we would like to extract a seed for a PRG (ChaCha20). We need a function to link both information, and achieve 2 properties:
  - the PRG seed must have a uniform entropy, and we can't assume the entropy source achieves that. We should then use a function to "smoothen" the entropy and uniformize it over all the bits. Such function can be a hash function, a MAC, KDF..
  - the PRG seed must be of some specific length (32 bytes). For better abstraction, we should not assume the entropy input has exactly that length. We can use a function with a variable output size, regardless of the input size. 
KDF, some MACs.. satisfy the properties above, I chose HKDF (a specific instance of KDF), but other functions are possible too. 

I'll improve the code doc to reflect the above. 

> as we seed on each block, what does this mean ?
 
PRGs have an internal state, that is initialized with a seed, and is updated each time it produces a random. If the state is known, all future outputs can be known. The state of some PRGs can be recovered based on querying multiple successive outputs. Imagine a transaction that doesn't know what seed is used but queries many randoms to recover the state, and predict the future outputs. This is not possible using a CSPRG like ChaCha20, i.e it's not possible to recover the state. 
Note that we eventually want to include the transaction index to seed the PRG per transaction, not per block.   

> considering number of randoms we can use in a transaction, is this important?

Probably not if the calls aren't many. I'm not sure how many calls a transaction can make, or maybe some applications can gather a large amount of randoms off-chain through multiple txs/blocks. 
Since we can output better quality randoms without extra cost, I think it's worth it to replace the lower quality one.